pkg/controllers/monitoring/monitoring-stack/components.go

```diff
@@ -41,27 +42,57 @@ func defaultReconciler(component client.Object, ms *stack.MonitoringStack) recon
+			if errors.IsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+		return nil
+	}
+}
+
+func alertManagerObjects(ms *stack.MonitoringStack, instanceSelectorKey string, instanceSelectorValue string) []client.Object {
+	alertmanagerRBACResourceName := ms.Name + "-alertmanager"
+	rbacVerbs := []string{"get", "list", "watch"}
+	return []client.Object{
+		newServiceAccount(alertmanagerRBACResourceName, ms.Namespace),
+		newAlertManagerRole(ms, alertmanagerRBACResourceName, rbacVerbs),
+		newRoleBinding(ms, alertmanagerRBACResourceName),
+		newAlertmanager(ms, alertmanagerRBACResourceName, instanceSelectorKey, instanceSelectorValue),
+		newAlertmanagerService(ms, instanceSelectorKey, instanceSelectorValue),
+		newAlertmanagerPDB(ms, instanceSelectorKey, instanceSelectorValue),
+	}
+}
+
 func stackComponentReconcilers(ms *stack.MonitoringStack, instanceSelectorKey string, instanceSelectorValue string) []reconcileFunction {
 	prometheusRBACResourceName := ms.Name + "-prometheus"
-	alertmanagerRBACResourceName := ms.Name + "-alertmanager"
 	rbacVerbs := []string{"get", "list", "watch"}
 	additionalScrapeConfigsSecretName := ms.Name + "-prometheus-additional-scrape-configs"
-	return []reconcileFunction{
+
+	reconcileFunctions := []reconcileFunction{
 		defaultReconciler(newServiceAccount(prometheusRBACResourceName, ms.Namespace), ms),
 		defaultReconciler(newPrometheusRole(ms, prometheusRBACResourceName, rbacVerbs), ms),
 		defaultReconciler(newRoleBinding(ms, prometheusRBACResourceName), ms),
 		defaultReconciler(newAdditionalScrapeConfigsSecret(ms, additionalScrapeConfigsSecretName), ms),
-		defaultReconciler(newServiceAccount(alertmanagerRBACResourceName, ms.Namespace), ms),
-		defaultReconciler(newAlertManagerRole(ms, alertmanagerRBACResourceName, rbacVerbs), ms),
-		defaultReconciler(newRoleBinding(ms, alertmanagerRBACResourceName), ms),
-		defaultReconciler(newAlertmanager(ms, alertmanagerRBACResourceName, instanceSelectorKey, instanceSelectorValue), ms),
-		defaultReconciler(newAlertmanagerService(ms, instanceSelectorKey, instanceSelectorValue), ms),
-		defaultReconciler(newAlertmanagerPDB(ms, instanceSelectorKey, instanceSelectorValue), ms),
 		defaultReconciler(newPrometheus(ms, prometheusRBACResourceName, additionalScrapeConfigsSecretName, instanceSelectorKey, instanceSelectorValue), ms),
 		defaultReconciler(newPrometheusService(ms, instanceSelectorKey, instanceSelectorValue), ms),
 		defaultReconciler(newThanosSidecarService(ms, instanceSelectorKey, instanceSelectorValue), ms),
 		prometheusPDBReconciler(newPrometheusPDB(ms, instanceSelectorKey, instanceSelectorValue), ms),
 	}
+
+	for _, amObj := range alertManagerObjects(ms, instanceSelectorKey, instanceSelectorValue) {
+		if ms.Spec.AlertmanagerConfig.Disabled {
+			reconcileFunctions = append(reconcileFunctions, removeReconciler(amObj, ms))
+		} else {
+			reconcileFunctions = append(reconcileFunctions, defaultReconciler(amObj, ms))
+		}
+	}
+
+	return reconcileFunctions
```

Comment:
Another option would be to have the reconcile function inspect `ms.Spec.AlertmanagerConfig.Disabled` similarly to what `prometheusPDBReconciler` is doing.
This would avoid treating the alertmanager resources differently in this location.
But that's nothing that prevents this from merging...

Reply:
Or instead of `prometheusPDBReconciler`, the controller could decide to use `defaultReconciler` (replicas > 1) or `removeReconciler`.

Reply:
Hmm so the suggestions are either have multiple functions and keep `stackComponentReconcilers` simple or have two functions (`defaultReconciler` and `removeReconciler` + possibly more in the futture) and have the logic in the `stackComponentReconcilers` right? 

I think both are possible...

Reply:
Just to offer a third version:
We could have a generic reconciler function that takes an additional argument: `removeReconciler(component client.Object, ms *stack.MonitoringStack, deleteCondition bool) reconcileFunction`.

I'm not the biggest fan of repeating blocks of
```
if condition1 {
		reconcileFunctions = append(reconcileFunctions, removeReconciler(obj1, ms))
} else {
	reconcileFunctions = append(reconcileFunctions, defaultReconciler(obj1, ms))
}
if condition2 {
		reconcileFunctions = append(reconcileFunctions, removeReconciler(obj2, ms))
} else {
	reconcileFunctions = append(reconcileFunctions, defaultReconciler(obj2, ms))
}
...
```
Is there and advantage of the above I'm not seeing?

Reply:
also to reiterate: I don't think this discussion needs to hold up this PR. I have a branch to put the reconcile logic in a module and use it in the thanos querier controller too.

Reply:
> Is there and advantage of the above I'm not seeing?

I don't think so. If you have it prepared in your branch then I would leave it.

Reply:
See #185 for what I had in mind. I'd very much appreciate feedback.

With that in mind, I'm happy to move this forward and discuss the reconcile bit on the other PR.

Reply:
I feel the code reads cleaner if the decision making part moves to a separate function that returns the right reconciler to use. 

e.g. the code would read as 


```go
reconcilers =  []reconcileFunction {
   defaultReconciler(...),
   pdbReconciler(ms), 
}

reconcilers = append(reconcilers, ...alertmanagerReconcilers(ms))
return reconcilers
```
where `pdbReconciler` will return the `remove|default` reconciler
and `alertmanagerReconcilers` will return the appropiate reconcilers

Reply:
So do you want me to update it? It seems unnecessary given the Jan's proposal (in different PR).

Reply:
I'm ok with leaving this as is.